File: src/twitch.ts
import ky, { HTTPError, type KyInstance } from "ky";
import { z } from "zod";
import { logger } from "./logger.ts";

const DEFAULT_HELIX_URL = "https://api.twitch.tv/helix";
const DEFAULT_TOKEN_URL = "https://id.twitch.tv/oauth2/token";

/** Max `user_login` values Twitch Helix accepts per `/streams` request. */
const MAX_LOGINS_PER_REQUEST = 100;

/** App Access Token (client-credentials) response. Only `access_token` is used. */
const tokenSchema = z
  .object({
    access_token: z.string(),
    expires_in: z.number().optional(),
    token_type: z.string().optional(),
  })
  .loose();

/**
 * `GET /helix/streams` response. A login is present in `data` only when the
 * channel is live, so the mere presence of an entry signals liveness.
 */
const streamsSchema = z
  .object({
    data: z.array(z.object({ user_login: z.string() }).loose()),
  })
  .loose();

function isUnauthorized(error: unknown): boolean {
  return error instanceof HTTPError && error.response.status === 401;
}

function chunk<T>(items: readonly T[], size: number): T[][] {
  const chunks: T[][] = [];
  for (let index = 0; index < items.length; index += size) {
    chunks.push(items.slice(index, index + size));
  }

  return chunks;
}

/**
 * Twitch Helix client. Obtains an App Access Token via the client-credentials
 * flow, caches it, and reports which of a set of logins are currently live.
 * The token lifecycle is fully internal — callers only ever see live logins.
 */
export class TwitchClient {
  readonly #helix: KyInstance;
  readonly #tokenUrl: string;
  readonly #clientId: string;
  readonly #clientSecret: string;
  #token: string | undefined;

  constructor(options: {
    clientId: string;
    clientSecret: string;
    helixUrl?: string;
    tokenUrl?: string;
  }) {
    this.#clientId = options.clientId;
    this.#clientSecret = options.clientSecret;
    this.#tokenUrl = options.tokenUrl ?? DEFAULT_TOKEN_URL;
    this.#helix = ky.create({
      prefix: options.helixUrl ?? DEFAULT_HELIX_URL,
      headers: {
        "Client-Id": options.clientId,
        Accept: "application/json",
      },
      retry: { limit: 2 },
      timeout: 10_000,
    });
  }

  /**
   * Returns the subset of `usernames` currently live on Twitch.
   *
   * Batches into Helix requests of up to 100 logins. Empty input performs no
   * HTTP at all (not even a token fetch). The App Access Token is fetched
   * lazily on first use and refreshed once on a `401`.
   */
  async fetchLiveUsernames(usernames: string[], signal?: AbortSignal): Promise<Set<string>> {
    const live = new Set<string>();
    if (usernames.length === 0) {
      return live;
    }

    let token = await this.#ensureToken();

    for (const logins of chunk(usernames, MAX_LOGINS_PER_REQUEST)) {
      let data: z.infer<typeof streamsSchema>["data"];
      try {
        data = await this.#queryStreams(logins, token, signal);
      } catch (error) {
        if (!isUnauthorized(error)) {
          throw error;
        }
        // Token likely expired: refresh once and retry this chunk.
        token = await this.#refreshToken();
        data = await this.#queryStreams(logins, token, signal);
      }

      for (const stream of data) {
        live.add(stream.user_login.toLowerCase());
      }
    }

    logger.debug(`Twitch reports ${live.size} live channel(s)`);

    return live;
  }

  async #queryStreams(
    logins: string[],
    token: string,
    signal?: AbortSignal,
  ): Promise<z.infer<typeof streamsSchema>["data"]> {
    const searchParams = new URLSearchParams(
      logins.map((login): [string, string] => ["user_login", login]),
    );
    const response = await this.#helix
      .get("streams", {
        searchParams,
        headers: { Authorization: `Bearer ${token}` },
        signal,
      })
      .json(streamsSchema);

    return response.data;
  }

  async #ensureToken(): Promise<string> {
    if (this.#token !== undefined) {
      return this.#token;
    }

    return this.#refreshToken();
  }

  async #refreshToken(): Promise<string> {
    const body = new URLSearchParams({
      client_id: this.#clientId,
      client_secret: this.#clientSecret,
      grant_type: "client_credentials",
    });
    const { access_token } = await ky
      .post(this.#tokenUrl, { body, retry: { limit: 2 }, timeout: 10_000 })
      .json(tokenSchema);
    this.#token = access_token;
    logger.debug("Twitch obtained an app access token");

    return access_token;
  }
}

File: internal/twitch/twitch.go
// Package twitch reports which of a set of Twitch logins are currently live,
// using the Helix API (github.com/nicklaw5/helix) with the client-credentials
// app-token flow.
package twitch

import (
	"context"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"

	helix "github.com/nicklaw5/helix/v2"

	"github.com/timche/teamspeak-stream-live/internal/logger"
)

const (
	// maxLoginsPerRequest is the Helix cap on `user_login` values per request.
	maxLoginsPerRequest = 100
	requestTimeout      = 10 * time.Second
)

// Options configures a Client. HTTPClient is optional and used by tests to route
// Helix and OAuth requests at a stub server.
type Options struct {
	ClientID     string
	ClientSecret string
	HTTPClient   *http.Client
}

// Client reports live Twitch logins. It wraps a helix.Client, which attaches the
// app access token to requests and transparently refreshes it on a 401 once an
// initial token is set — so this wrapper only fetches the first token and batches
// the logins.
type Client struct {
	helix *helix.Client

	mu        sync.Mutex
	haveToken bool
}

// New builds a Client.
func New(opts Options) (*Client, error) {
	httpClient := opts.HTTPClient
	if httpClient == nil {
		httpClient = &http.Client{Timeout: requestTimeout}
	}
	h, err := helix.NewClient(&helix.Options{
		ClientID:     opts.ClientID,
		ClientSecret: opts.ClientSecret,
		HTTPClient:   httpClient,
	})
	if err != nil {
		return nil, err
	}
	return &Client{helix: h}, nil
}

// FetchLiveUsernames returns the subset of usernames currently live on Twitch.
// It batches into Helix requests of up to 100 logins and performs no HTTP at all
// on empty input. The ctx is part of the interface for parity with the other
// source; helix binds its context at construction, so cancellation is bounded by
// the HTTP client timeout rather than this ctx.
func (c *Client) FetchLiveUsernames(_ context.Context, usernames []string) (map[string]struct{}, error) {
	live := make(map[string]struct{})
	if len(usernames) == 0 {
		return live, nil
	}

	if err := c.ensureToken(); err != nil {
		return nil, err
	}

	for _, logins := range chunk(usernames, maxLoginsPerRequest) {
		// First defaults to 20 server-side, so it is set to the batch size to
		// return every live channel in the batch on one page. (The original
		// TypeScript omitted it and silently reported only the first 20.)
		resp, err := c.helix.GetStreams(&helix.StreamsParams{
			UserLogins: logins,
			First:      maxLoginsPerRequest,
		})
		if err != nil {
			return nil, err
		}
		if resp.StatusCode != http.StatusOK {
			return nil, fmt.Errorf("twitch streams request failed: %d %s", resp.StatusCode, resp.ErrorMessage)
		}
		for _, stream := range resp.Data.Streams {
			live[strings.ToLower(stream.UserLogin)] = struct{}{}
		}
	}

	logger.Log.Debug("Twitch live channels", "count", len(live))
	return live, nil
}

// ensureToken fetches the initial app access token once (retrying on a later
// poll if it fails). Setting the token also enables helix's built-in refresh on
// a subsequent 401.
func (c *Client) ensureToken() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.haveToken {
		return nil
	}

	resp, err := c.helix.RequestAppAccessToken(nil)
	if err != nil {
		return err
	}
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("twitch token request failed: %d %s", resp.StatusCode, resp.ErrorMessage)
	}

	c.helix.SetAppAccessToken(resp.Data.AccessToken)
	c.haveToken = true
	logger.Log.Debug("Twitch obtained an app access token")
	return nil
}

func chunk[T any](items []T, size int) [][]T {
	var chunks [][]T
	for start := 0; start < len(items); start += size {
		end := start + size
		if end > len(items) {
			end = len(items)
		}
		chunks = append(chunks, items[start:end])
	}
	return chunks
}
